### PR TITLE
ADM-924: [frontend] could export multiple sets of data in report list page

### PR DESCRIPTION
**Pull request by TingyuDong**
## Summary

...

## Before

image

**Screenshots**

## After

image

## Note

_Null_

### CODE REVIEW

File: frontend/__tests__/containers/ReportButtonGroup.test.tsx [hunk 1]
```diff
@@ -36,18 +126,129 @@ describe('test', () => {
...
+
+  it('should export data buttons be not clickable when the CSV file for any dataRange has not been generated', () => {
+    setup(pendingMockData);
+
+    expect(screen.getByRole('button', { name: EXPORT_METRIC_DATA })).toBeDisabled();
+    expect(screen.getByRole('button', { name: EXPORT_BOARD_DATA })).toBeDisabled();
+    expect(screen.getByRole('button', { name: EXPORT_PIPELINE_DATA })).toBeDisabled();
+  });
+
+  [EXPORT_METRIC_DATA, EXPORT_BOARD_DATA, EXPORT_PIPELINE_DATA].forEach((buttonName) => {
```
Comment (Leiqiuhong) [inline frontend/__tests__/containers/ReportButtonGroup.test.tsx:191]: Let's change it to it.each

File: frontend/__tests__/containers/ReportButtonGroup.test.tsx [hunk 2]
```diff
@@ -36,18 +126,129 @@ describe('test', () => {
...
+    expect(checkbox).not.toBeDisabled();
+    await userEvent.click(checkbox);
+
+    await userEvent.click(screen.getByText('Confirm'));
+
+    expect(screen.queryByText('Select the time period for the exporting data')).not.toBeInTheDocument();
+    expect(result.current.fetchExportData).toBeCalledTimes(1);
+  });
+
+  [EXPORT_METRIC_DATA, EXPORT_BOARD_DATA, EXPORT_PIPELINE_DATA].forEach((buttonName) => {
```
Comment (Leiqiuhong) [inline frontend/__tests__/containers/ReportButtonGroup.test.tsx:245]: Let's change it to it.each

**Comment by guzhongren** [comment]
1 sonar issue: https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&pullRequest=1436&id=au-heartbeat-heartbeat-frontend&open=AY9ce38RQA-dYWUTFPTq

File: frontend/__tests__/containers/ReportButtonGroup.test.tsx [hunk 3]
```diff
@@ -1,31 +1,121 @@
...
+    boardMetricsCompleted: true,
+    doraMetricsCompleted: true,
     reportMetricsError: {
-      boardMetricsError: {
-        status: 401,
-        message: 'Unauthorized',
+      boardMetricsError: null,
+      pipelineMetricsError: null,
+      sourceControlMetricsError: null,
+    },
```
Comment (guzhongren) [inline frontend/__tests__/containers/ReportButtonGroup.test.tsx:45]: extract this part as a obj, then ...obj

File: frontend/__tests__/containers/ReportButtonGroup.test.tsx [hunk 4]
```diff
@@ -1,31 +1,121 @@
...
+    {
+      ...secondBasicMockDateRangeRequestResult,
+      reportMetricsError: {
+        boardMetricsError: {
+          status: 500,
+          message: 'mockError',
+        },
+        pipelineMetricsError: {
+          status: 500,
+          message: 'mockError',
```
Comment (guzhongren) [inline frontend/__tests__/containers/ReportButtonGroup.test.tsx:108]: duplicated code. same as previous one

File: frontend/__tests__/containers/ReportButtonGroup.test.tsx [hunk 5]
```diff
@@ -36,18 +126,128 @@ describe('test', () => {
...
+          csvTimeStamp={1715250961572}
+          dateRangeRequestResults={dateRangeRequestResults}
         />
       </Provider>,
     );
   };
 
-  it('test', () => {
-    setup();
+  it('should all buttons be clickable when the request successfully finishes', () => {
```
Comment (guzhongren) [inline frontend/__tests__/containers/ReportButtonGroup.test.tsx:136]: `when` --> `given`

File: frontend/__tests__/containers/ReportButtonGroup.test.tsx [hunk 6]
```diff
@@ -36,18 +126,128 @@ describe('test', () => {
...
+    expect(checkbox).not.toBeDisabled();
+    await userEvent.click(checkbox);
+
+    await userEvent.click(screen.getByText('Confirm'));
+
+    expect(screen.queryByText('Select the time period for the exporting data')).not.toBeInTheDocument();
+    expect(result.current.fetchExportData).toBeCalledTimes(1);
+  });
+
+  it(`should should not be able to click the button when an error occurs for all dataRanges`, async () => {
```
Comment (guzhongren) [inline frontend/__tests__/containers/ReportButtonGroup.test.tsx:246]: `should should` what are u doing?

File: frontend/__tests__/containers/ReportStep/DownloadDialog.test.tsx [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    setup(mockData);
+
+    expect(screen.getByText('2024/01/01 - 2024/01/14')).toBeInTheDocument();
+    expect(screen.getByText('2024/01/15 - 2024/01/31')).toBeInTheDocument();
+    expect(screen.getAllByRole('checkbox')[0]).not.toBeDisabled();
+    expect(screen.getAllByRole('checkbox')[1]).not.toBeDisabled();
+    expect(screen.getByText('Confirm')).toBeDisabled();
+  });
+
+  it('should not be clickable when there is an error fetching data for this dataRange', () => {
```
Comment (guzhongren) [inline frontend/__tests__/containers/ReportStep/DownloadDialog.test.tsx:46]: change `when` to `given` when necessary,
firstly, you need to define the `when` and `given` in test description

File: frontend/__tests__/containers/ReportStep/DownloadDialog.test.tsx [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    const confirmButton = screen.getByText('Confirm');
+    expect(confirmButton).not.toBeDisabled();
+
+    await userEvent.click(confirmButton);
+
+    expect(handleCloseFn).toBeCalledTimes(1);
+    expect(downloadCSVFileFn).toBeCalledTimes(1);
+  });
+
+  it('should not choose the dataRange when click on the already selected dataRange', async () => {
```
Comment (guzhongren) [inline frontend/__tests__/containers/ReportStep/DownloadDialog.test.tsx:96]: when 后面的动词应该是ing 形式

File: frontend/src/theme.ts [hunk 9]
```diff
@@ -143,6 +146,9 @@ export const theme = createTheme({
         borderColor: '#939DDA',
       },
     },
+    downloadListLabel: {
+      backgroundColor: 'rgba(67, 80, 175, 0.1)',
```
Comment (guzhongren) [inline frontend/src/theme.ts:150]: covert it to `#....`

**Review by guzhongren — APPROVED** [review]
code changes

**Review by PengxiWPix — APPROVED** [review]
LGTM

File: frontend/__tests__/containers/ReportButtonGroup.test.tsx (hunk 1 above)
Comment (TingyuDong) [reply frontend/__tests__/containers/ReportButtonGroup.test.tsx]: Code has been changed.

File: frontend/__tests__/containers/ReportButtonGroup.test.tsx (hunk 2 above)
Comment (TingyuDong) [reply frontend/__tests__/containers/ReportButtonGroup.test.tsx]: Code has been changed.

File: frontend/__tests__/containers/ReportButtonGroup.test.tsx (hunk 3 above)
Comment (TingyuDong) [reply frontend/__tests__/containers/ReportButtonGroup.test.tsx]: Code has been changed.

File: frontend/__tests__/containers/ReportButtonGroup.test.tsx (hunk 5 above)
Comment (TingyuDong) [reply frontend/__tests__/containers/ReportButtonGroup.test.tsx]: Code has been changed.

File: frontend/__tests__/containers/ReportButtonGroup.test.tsx (hunk 6 above)
Comment (TingyuDong) [reply frontend/__tests__/containers/ReportButtonGroup.test.tsx]: Code has been changed.

File: frontend/__tests__/containers/ReportStep/DownloadDialog.test.tsx (hunk 7 above)
Comment (TingyuDong) [reply frontend/__tests__/containers/ReportStep/DownloadDialog.test.tsx]: Code has been changed.

File: frontend/__tests__/containers/ReportStep/DownloadDialog.test.tsx (hunk 8 above)
Comment (TingyuDong) [reply frontend/__tests__/containers/ReportStep/DownloadDialog.test.tsx]: Code has been changed.

File: frontend/src/theme.ts (hunk 9 above)
Comment (TingyuDong) [reply frontend/src/theme.ts]: Code has been changed.

File: frontend/src/containers/ReportStep/index.tsx [hunk 10]
```diff
@@ -111,6 +121,23 @@ const ReportStep = ({ handleSave }: ReportStepProps) => {
...
 
+  const mapDateResult = (descendingDateRanges: DateRange, reportInfos: IReportInfo[]) =>
+    descendingDateRanges.map(({ startDate, endDate }) => {
+      const reportData = reportInfos.find((singleResult) => singleResult.id === startDate)?.reportData ?? null;
+      return {
+        startDate: startDate,
+        endDate: endDate,
+        overallMetricsCompleted: reportData?.overallMetricsCompleted ?? false,
+        boardMetricsCompleted: reportData?.boardMetricsCompleted ?? null,
+        doraMetricsCompleted: reportData?.doraMetricsCompleted ?? null,
```
Comment (mrcuriosity-tw) [inline frontend/src/containers/ReportStep/index.tsx:132]: `overallMetricsCompleted` potentially infers to a `Boolean`, however `boardMetricsCompleted` & `doraMetricsCompleted` infer to `Nullable<boolean>`. How about using a same type?

**Review by mrcuriosity-tw — APPROVED** [review]

**Pull request merged**